inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationUpdateService.java

```diff
@@ -0,0 +1,176 @@
+        val newInstrumentation = ev.getNewConfig().getInstrumentation();
+
+        //if the instrumentation configuration has changed all classes need to be reevalauted
+        if (!Objects.equals(oldInstrumentation, newInstrumentation)) {
+            requestConfigUpdateForAllLoadedClasses();
+        }
+
+        InternalSettings newInternal = newInstrumentation.getInternal();
+        classCheckerJob.setBatchSizes(newInternal.getClassConfigurationCheckBatchSize());
+        classCheckerJob.setInterBatchDelay(newInternal.getClassConfigurationCheckInterBatchDelay());
+    }
+
+
+    /**
+     * Takes numberOfClassesToProcess classes from {@link #pendingClasses} and calls {@link #ensureInstrumentationIsUpToDate(Class)}
+     * for them. This method is invoked regularly by {@link #classCheckerJob}.
+     * <p>
+     * This method is package-private for testing.
+     *
+     * @param numberOfClassesToProcess the number of classes to take from {@link #pendingClasses}
+     */
+    void checkClassesForConfigurationUpdates(int numberOfClassesToProcess) {
+        if (transformer.isShuttingDown()) {
+            return; //nothing to do anymore
+        }
+
+        val watch = new StopWatch();
+        try {
+            int processedClassesCount;
+            for (processedClassesCount = 0; processedClassesCount < numberOfClassesToProcess; processedClassesCount++) {
+                Optional<Class<?>> nextClass = CommonUtils.pollElementFromSet(pendingClasses);
+                if (nextClass.isPresent()) {
+                    pendingClasses.remove(nextClass.get());
+                    ensureInstrumentationIsUpToDate(nextClass.get());
+                } else {
+                    break;
+                }
+            }
+            if (processedClassesCount > 0) {
+                log.debug("Checked configuration of {} classes in {} ms, {} classes left to check", processedClassesCount, watch.getElapsedMillis(), pendingClasses.size());
+            }
+        } catch (Exception e) {
+            log.error("Error checking for class instrumentation configuration updates", e);
+        }
+    }
+
+    /**
+     * Checks if the active instrumentation for the given class matches the one requested through the configuration.
+     * If the class is not up to date it issued to the {@link #retransformationService}
+     *
+     * @param clazz the class to check.
+     */
+    private void ensureInstrumentationIsUpToDate(Class<?> clazz) {
+        if (!transformer.isIgnoredClass(clazz) && doesClassRequireRetransformation(clazz)) {
+            log.debug("Detected config change for {}, queing for retransform", clazz.getName());
+            retransformationService.requestRetransformation(clazz);
+        }
+    }
+
+    /**
+     * Todo: where to actually put this method... ? (Code duplication)
+     *
+     * @param clazz
+     * @return
+     */
+    private boolean doesClassRequireRetransformation(Class<?> clazz) {
```

Comment:
we are doing the same thing twice, first here, then in the retransform service.. Not really sure if this is the part where you optimize for this situation where you can quickly change instrumentation twice.. If this is the case, I would remove this optimization which would also make the `ClassRetransformationService` actually not needed as we can request the instrumentation directly from here.. It would decrease the complexity, speed things up and the only downside is covering the situation that happens very very rare.. Let's please discuss this..